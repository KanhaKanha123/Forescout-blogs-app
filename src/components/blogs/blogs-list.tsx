import { FC, ReactElement } from "react";
import { Button, Card, HandleLoadingEmptyErrorState } from "../shared-comonents";
import {
    BlogsWrapper,
    Image,
    Text,
    CommentsTextAndButtonWrapper,
    Header,
    ReactionsContainer,
    TagsContainer,
    BodyContainer,
    TitleTextWrapper,
    BlogBody
} from "./style";
import chatImg from "../../assets/icons/chat-icon.jpeg";
import { useNavigate } from 'react-router-dom';
import { useGetBlogs } from "../../custom-hooks";

/**
 * This component is responsible to render all the blogs in cards with few details.
 * @returns ReactElement
 */
export const BlogsList: FC = (): ReactElement => {
    const navigate = useNavigate();

    //get blogs data from custom hook
    const { blogs, isLoading, error } = useGetBlogs();

    //on click navigate to details page
    const cardClickHandler = (blogId: number) => {
        navigate(`/blog/${blogId}`, { replace: true });
    }

    return (
        <HandleLoadingEmptyErrorState data={blogs} isLoading={isLoading} error={error}>
            <BlogsWrapper>
                {(blogs || []).map(post => <Card
                    data-testid={`Card_${post.id}`}
                    aria-label="card containing the blog data"
                    key={post.id}
                    width="300px"
                    height="200px"
                    padding="15px"
                    backgroundcolor="#f4f4f4"
                    onClick={() => cardClickHandler(post.id)}>
                    <Header>
                        <TitleTextWrapper aria-label="blog title">
                            {post.title}
                        </TitleTextWrapper>
                    </Header>
                    <TagsContainer aria-label="blog tags listed here">
                        {post.tags.map((tag: string, index: number) => <Card
                            key={index}
                            width="70px"
                            height="25px"
                            backgroundcolor="#8fbaf0">
                            {tag}
                        </Card>)}
                    </TagsContainer>
                    <BodyContainer aria-label="blog description is here">
                        <BlogBody>{post.body.substring(0, 50)}...</BlogBody>
                    </BodyContainer>
                    <CommentsTextAndButtonWrapper aria-label="blog reactions count is here">
                        <ReactionsContainer>
                            <Image aria-label="icon to show reaction" src={chatImg}></Image>
                            <Text aria-label="reaction text here">
                                {post.reactions}
                            </Text>
                        </ReactionsContainer>
                        <Button aria-label="read more button" backgroundcolor="#4792ef">Read More</Button>
                    </CommentsTextAndButtonWrapper>
                </Card>)}
            </BlogsWrapper>
        </HandleLoadingEmptyErrorState>);
};