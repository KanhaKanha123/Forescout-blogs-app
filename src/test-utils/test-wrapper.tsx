import ContextProvider from '../store/context';
import { QueryClientProvider } from "react-query";
import { FC, ReactElement } from 'react';
import { client } from '../react-query/query-client';
import { BrowserRouter as Router } from 'react-router-dom';

interface TestWrapperProps {
    children: React.ReactNode
}

/**
 * This component is responsible to wrap the tests with all the important wrappers so all the time in all tests we don't have to repeat.
 * @returns ReactElement
 */
export const TestWrapper: FC<TestWrapperProps> = ({ children }): ReactElement => (
    <QueryClientProvider client={client}>
        <ContextProvider>
            <Router>
                {children}
            </Router>
        </ContextProvider>
    </QueryClientProvider>);