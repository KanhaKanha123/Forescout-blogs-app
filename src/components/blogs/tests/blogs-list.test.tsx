import { fireEvent, render } from '@testing-library/react';
import { BlogsList } from '../blogs-list';
import { TestWrapper } from '../../../test-utils/test-wrapper';
import * as router from 'react-router';

//mock custom hook response
jest.mock('../../../custom-hooks/use-get-blogs', () => ({
  useGetBlogs: () => ({
    blogs: [{
      body: 'hello i am here',
      id: 1,
      reactions: 5,
      tags: ['one', 'two'],
      title: 'the best blog',
      userId: 1
    }],
    isLoading: false,
    error: null
  })
}));


jest.mock('../../../custom-hooks/use-get-user-data', () => ({
  useGetUserData: () => ({
    userData: {
      id: 1,
      firstName: 'test',
      lastName: 'test name',
      age: 30,
    }
  })
}));

const rederComponent = () => render(<TestWrapper>
  <BlogsList />
</TestWrapper>);

describe('BlogsList component', () => {

  const navigate = jest.fn()

  beforeEach(() => {
    jest.spyOn(router, 'useNavigate').mockImplementation(() => navigate)
  })

  test('renders BlogsList component proper', () => {

    const component = rederComponent();

    expect(component).toMatchSnapshot();
  });

  test('renders BlogsList component should have proper props', () => {

    const component = rederComponent();

    const blogBody = component.getByText(/hello i am here/i);

    const blogTitle = component.getByText(/the best blog/i);

    expect(blogBody).toBeInTheDocument();

    expect(blogTitle).toBeInTheDocument();
  });

  test('cardClickHandler should be called', () => {

    const component = rederComponent();

    const card = component.getByTestId('Card_1')

    fireEvent.click(card)

    expect(navigate).toHaveBeenCalledWith('/blog/1', { "replace": true })
  });

});
