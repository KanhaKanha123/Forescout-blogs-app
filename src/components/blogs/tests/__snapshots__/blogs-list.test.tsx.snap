// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlogsList component renders BlogsList component proper 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-gKkivs bgfSMQ"
      >
        <div
          aria-label="card containing the blog data"
          backgroundcolor="#f4f4f4"
          class="sc-bdDrbm fvGOqo"
          data-testid="Card_1"
          height="200px"
          padding="15px"
          width="300px"
        >
          <div
            class="sc-iCEALv dDRKGZ"
          >
            <span
              aria-label="blog title"
              class="sc-crPhKe QNeik"
            >
              the best blog
            </span>
          </div>
          <div
            aria-label="blog tags listed here"
            class="sc-fuzrJS lieATZ"
          >
            <div
              backgroundcolor="#8fbaf0"
              class="sc-bdDrbm cXPPEs"
              height="25px"
              width="70px"
            >
              one
            </div>
            <div
              backgroundcolor="#8fbaf0"
              class="sc-bdDrbm cXPPEs"
              height="25px"
              width="70px"
            >
              two
            </div>
          </div>
          <div
            aria-label="blog description is here"
            class="sc-qdPnm fmmVKh"
          >
            <p
              class="sc-bjLqKs hcrKFR"
            >
              hello i am here
              ...
            </p>
          </div>
          <div
            aria-label="blog reactions count is here"
            class="sc-jrICsz hYBaNZ"
          >
            <div
              class="sc-dYlqv cvZwVn"
            >
              <img
                aria-label="icon to show reaction"
                class="sc-iqkjcl jA-dEtT"
                src="chat-icon.jpeg"
              />
              <span
                aria-label="reaction text here"
                class="sc-folRzf jAzsXY"
              >
                5
              </span>
            </div>
            <div
              aria-label="read more button"
              backgroundcolor="#4792ef"
              class="sc-gtcAbF dPqWXl"
            >
              Read More
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-gKkivs bgfSMQ"
    >
      <div
        aria-label="card containing the blog data"
        backgroundcolor="#f4f4f4"
        class="sc-bdDrbm fvGOqo"
        data-testid="Card_1"
        height="200px"
        padding="15px"
        width="300px"
      >
        <div
          class="sc-iCEALv dDRKGZ"
        >
          <span
            aria-label="blog title"
            class="sc-crPhKe QNeik"
          >
            the best blog
          </span>
        </div>
        <div
          aria-label="blog tags listed here"
          class="sc-fuzrJS lieATZ"
        >
          <div
            backgroundcolor="#8fbaf0"
            class="sc-bdDrbm cXPPEs"
            height="25px"
            width="70px"
          >
            one
          </div>
          <div
            backgroundcolor="#8fbaf0"
            class="sc-bdDrbm cXPPEs"
            height="25px"
            width="70px"
          >
            two
          </div>
        </div>
        <div
          aria-label="blog description is here"
          class="sc-qdPnm fmmVKh"
        >
          <p
            class="sc-bjLqKs hcrKFR"
          >
            hello i am here
            ...
          </p>
        </div>
        <div
          aria-label="blog reactions count is here"
          class="sc-jrICsz hYBaNZ"
        >
          <div
            class="sc-dYlqv cvZwVn"
          >
            <img
              aria-label="icon to show reaction"
              class="sc-iqkjcl jA-dEtT"
              src="chat-icon.jpeg"
            />
            <span
              aria-label="reaction text here"
              class="sc-folRzf jAzsXY"
            >
              5
            </span>
          </div>
          <div
            aria-label="read more button"
            backgroundcolor="#4792ef"
            class="sc-gtcAbF dPqWXl"
          >
            Read More
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
